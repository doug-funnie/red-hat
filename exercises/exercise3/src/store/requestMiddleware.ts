import { Middleware } from 'redux';
import { Action } from 'redux-actions';

export enum Lifecycle {
  requested,
  success,
  error,
}

interface RequestAction<SP = any, RP = undefined> {
  type: string;
  request(a: RP): Promise<SP>;
  payload: RP;
  meta?: any;
}

export interface RequestMiddlewareAction<P = any> {
  type: string;
  payload: P;
  meta: {
    'request-middleware': Lifecycle;
    [key: string]: any;
  };
  error?: boolean;
}

// forces typings to match
type RequestMiddlewareKey = keyof Pick<RequestMiddlewareAction['meta'], 'request-middleware'>;
export const requestMiddlewareKey: RequestMiddlewareKey = 'request-middleware';

const isRequestAction = (action: RequestAction): action is RequestAction => !!action.request;
const isRequestMiddlewareAction = (action: RequestMiddlewareAction): action is RequestMiddlewareAction => (
  action.meta && action.meta[requestMiddlewareKey] !== null
);

export function createRequestAction<SP>(type: string, request: () => Promise<SP>): () => RequestAction<SP>;
export function createRequestAction<RP, SP>(
  type: string,
  request: (requestPayload: RP) => Promise<SP>,
): (a: RP) => RequestAction<SP, RP>;
export function createRequestAction(type, request) {
  return (a): RequestAction => ({
    type,
    request,
    payload: a,
  });
}

const requestHasExpired = (lastRequestedTime: number) => {
  if (!lastRequestedTime) {
    return true;
  }
  const diff = performance.now() - lastRequestedTime;
  return diff >= 300000; // 5 Minutes
};

export const createRequestMiddleware = () => {
  const requestCache: {
    [url: string]: {
      request: Promise<any>;
      lastRequestedTime: number;
    };
  } = {};
  const cacheRequest = (type: string, request: Promise<any>) => {
    requestCache[type] = {
      request,
      lastRequestedTime: performance.now(),
    };
  };

  const middleware: Middleware = ({ dispatch }) => (next) => async (action) => {
    if (!isRequestAction(action)) {
      return next(action);
    }

    const cachedRequest = requestCache[action.type];
    if (cachedRequest && cachedRequest.request && !requestHasExpired(cachedRequest.lastRequestedTime)) {
      return cachedRequest.request;
    }

    const createRequestMiddlewareAction = (lifecycle: Lifecycle, payload: any): RequestMiddlewareAction => ({
      payload,
      type: action.type,
      error: lifecycle === Lifecycle.error,
      meta: {
        ...action.meta,
        [requestMiddlewareKey]: lifecycle,
      },
    });

    dispatch(createRequestMiddlewareAction(Lifecycle.requested, action.payload));
    try {
      const request = action.request(action.payload);
      cacheRequest(action.type, request);
      const data = await(request);
      dispatch(createRequestMiddlewareAction(Lifecycle.success, data));
      return data;
    } catch (err) {
      dispatch(createRequestMiddlewareAction(Lifecycle.error, err));
      return Promise.reject(err);
    }
  };

  return middleware;
};

type RequestActionHandler<S = any, P = any> = (prevState: S, action: Action<P>) => S;

interface RequestActionHandlers<S, RP, SP, EP> {
  requested?: RequestActionHandler<S, RP>;
  success?: RequestActionHandler<S, SP>;
  error?: RequestActionHandler<S, EP>;
  finally?: RequestActionHandler<S, RP | EP>;
}

export function handleRequestAction<S, SP, RP = any, EP = any>(
  type: string,
  handlers: RequestActionHandlers<S, RP, SP, EP>,
  defaultState: S,
) {
  return (state: S = defaultState, action: RequestMiddlewareAction) => {
    if (action.type !== type) {
      return state;
    }

    if (!isRequestMiddlewareAction(action)) {
      return state;
    }

    const lifecycle = action.meta[requestMiddlewareKey];
    const handler = handlers[Lifecycle[lifecycle]];
    if (handler) {
      return handler(state, action);
    }
    return state;
  };
}
