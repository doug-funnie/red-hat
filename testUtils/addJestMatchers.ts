// import serializer from 'jest-glamor-react';
import { toMatchSnapshot } from 'jest-snapshot';

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeCalledWithSnapshot(snapshotName?: string): void;
    }
  }
}

// expect.addSnapshotSerializer(serializer);

expect.extend({
  toBeCalledWithSnapshot(received, snapshotName) {
    if (received.mock.calls.length === 0) {
      return {
        message: () => 'Mock function was not called',
        pass: false,
      };
    }

    return toMatchSnapshot.call(this, received.mock.calls, snapshotName);
  },
});
