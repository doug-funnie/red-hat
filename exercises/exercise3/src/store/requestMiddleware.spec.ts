import '../../../../testUtils/addJestMatchers';
import * as requestMiddleware from './requestMiddleware';
import { createRequestAction, RequestMiddlewareAction } from './requestMiddleware';

it('createRequestAction returns a request action', () => {
  const request = (_: string) => Promise.resolve('');
  const type = 'REQUEST_ACTION';
  const payload = 'testPayload';
  const actionCreator = requestMiddleware.createRequestAction(type, request);
  const action = actionCreator(payload);
  expect(action.type).toBe(type);
  expect(action.request).toBe(request);
  expect(action.payload).toBe(payload);
});

describe('middleware', () => {
  let middleware = requestMiddleware.createRequestMiddleware();
  const mockDispatch = jest.fn();
  const mockStore: any = {
    dispatch: mockDispatch,
  };
  const mockNext = jest.fn();
  const mockRequest = jest.fn(() => Promise.resolve(''));
  const mockRequestAction = createRequestAction('REQUEST_ACTION', mockRequest)();
  const mockAction = { type: 'ACTION' };

  beforeEach(() => {
    mockRequest.mockReturnValue(Promise.resolve('RESOLVED'));
    middleware = requestMiddleware.createRequestMiddleware();
  });

  it('calls next for actions that are not request actions', () => {
    middleware(mockStore)(mockNext)(mockAction);
    expect(mockNext).toBeCalledWith(mockAction);
  });

  it('dispatches success action', async () => {
    await middleware(mockStore)(mockNext)(mockRequestAction);
    expect(mockDispatch).toBeCalledWithSnapshot();
  });

  it('dispatches error action', async () => {
    try {
      mockRequest.mockReturnValue(Promise.reject('REJECTION'));
      await middleware(mockStore)(mockNext)(mockRequestAction);
    } catch(_) {} //tslint:disable-line
    expect(mockDispatch).toBeCalledWithSnapshot();
  });

  it('returns cached request if is within expiration time', async () => {
    const first = await middleware(mockStore)(mockNext)(mockRequestAction);
    const second = await middleware(mockStore)(mockNext)(mockRequestAction);
    expect(first).toBe(second);
  });
});

describe('handleRequestAction', () => {
  const type = 'ACTION_TYPE';
  const requestedHandler = jest.fn();
  const successHandler = jest.fn();
  const errorHandler = jest.fn();
  const defaultState = {};
  const handler = requestMiddleware.handleRequestAction(type, {
    error: errorHandler,
    success: successHandler,
    requested: requestedHandler,
  }, defaultState);

  const getAction = (lifecycle: requestMiddleware.Lifecycle): RequestMiddlewareAction => ({
    type,
    payload: '',
    meta: {
      'request-middleware': lifecycle,
    },
  });

  it('defaults state', () => {
    const nextState = handler(undefined, {} as any);
    expect(nextState).toBe(defaultState);
  });

  it('ignores actions with non-matching type', () => {
    const nextState = handler(undefined, { type: type + 'OTHER' } as any);
    expect(nextState).toBe(defaultState);
  });

  it('calls requested handler', () => {
    const action = getAction(requestMiddleware.Lifecycle.requested);
    handler(defaultState, action);
    expect(requestedHandler).toBeCalledWith(defaultState, action);
  });

  it('calls success handler', () => {
    const action = getAction(requestMiddleware.Lifecycle.success);
    handler(defaultState, action);
    expect(successHandler).toBeCalledWith(defaultState, action);
  });

  it('calls error handler', () => {
    const action = getAction(requestMiddleware.Lifecycle.error);
    handler(defaultState, action);
    expect(errorHandler).toBeCalledWith(defaultState, action);
  });
});
