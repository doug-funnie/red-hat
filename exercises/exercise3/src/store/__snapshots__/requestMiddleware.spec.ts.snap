// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware dispatches error action 1`] = `
Array [
  Array [
    Object {
      "error": false,
      "meta": Object {
        "request-middleware": 0,
      },
      "payload": undefined,
      "type": "REQUEST_ACTION",
    },
  ],
  Array [
    Object {
      "error": true,
      "meta": Object {
        "request-middleware": 2,
      },
      "payload": "REJECTION",
      "type": "REQUEST_ACTION",
    },
  ],
]
`;

exports[`middleware dispatches success action 1`] = `
Array [
  Array [
    Object {
      "error": false,
      "meta": Object {
        "request-middleware": 0,
      },
      "payload": undefined,
      "type": "REQUEST_ACTION",
    },
  ],
  Array [
    Object {
      "error": false,
      "meta": Object {
        "request-middleware": 1,
      },
      "payload": "RESOLVED",
      "type": "REQUEST_ACTION",
    },
  ],
]
`;
